Speed up the application of PiecewiseAffineTransform

svepe opened the pull request:
The speedup is achieved by replacing manual iteration over individual affine transforms with `np.einsum`.

Closes #6864

svepe committed:
Speed up the application of PiecewiseAffineTransform

Replace manual iteration over individual affine transforms with np.einsum

23pointsNorth commented:
Any updates on this? @lagru @stefanv

lagru commented:
Hey, sorry for our delayed response on this. I'll have a look.

lagru committed:
Merge branch 'main' into piecewise-affine-speedup

lagru commented:
I am noticing that for the old and new implementation the number seems to change a bit. E.g. for the example below the residuals change on a scale <1e-15.
```python
import skimage as ski
src = np.array([[0, 3], [0, -1], [1, 2], [-1, 0], [1, 1], [-1, 1]])
dst = np.array([[0, 0], [1, 0], [1, 1], [1, 2], [0, 2], [0, 1]])
tform = ski.transform.PiecewiseAffineTransform()
tform.estimate(src, dst)
tform.residuals(src, dst)
```
Probably floating point errors but not sure yet? I don't like that our tests don't catch this.

While we are at it, would you mind adapting the above example with outputs as a doctest? :)

svepe committed:
Add a doctest to check if PiecewiseAffine residuals change

svepe commented:
I've added the doctest to the `estimate` function.

lagru committed:
Use ">>>" in docstring

and try to address failure due to

"PiecewiseAffineTransform.estimate:58:Unexpected indentation." 

See https://dev.azure.com/scikit-image/scikit-image/_build/results?buildId=10795&view=logs&j=feb0b66f-d63c-5533-bf34-a72e08b12053&t=26b7b1f7-60f2-5676-21d5-45b533330524&l=1050

lagru commented:
Thanks @svepe! I tried to address the test failure in https://github.com/scikit-image/scikit-image/pull/6963/commits/949cf8a3b12e2d729ed658b90f43986ecdff2cdb.

lagru committed:
Relax doctest for residuals

svepe commented:
@lagru Let me know if there is anything else needed on my end to merge the PR.

lagru commented:
@svepe, I think this looks good. However, I am hesitant to approve this because this performance optimization does change our results, if only slightly. We have a policy to to never silently change the results if input stays the same. And because we don't seem to have a good test coverage for this, I am hesitant to assume from a single doctest that this is only noise due to floating errors.

Let's see what @scikit-image/core thinks and I'll bring it up in the next meeting as well.

lagru commented:
Point in favor of merging this is that the snippet below seems to pass always. We use `assert_allclose` elsewhere to check our output with a relative tolerance of 1e-7. So we might consider this change just noise.

<details><summary>Details</summary>
<p>

```python
import numpy as np
import skimage as ski

class NewPiecewiseAffineTransform(ski.transform.PiecewiseAffineTransform):
    def __call__(self, coords):
        coords = np.asarray(coords)
        # determine triangle index for each coordinate
        simplex = self._tesselation.find_simplex(coords)
        # stack of affine transforms to be applied to every coord
        affines = np.stack([affine.params for affine in self.affines])[simplex]
        # convert coords to homgeneous points
        points = np.c_[coords, np.ones((coords.shape[0], 1))]
        # apply affine transform to every point
        result = np.einsum("ikj,ij->ik", affines, points)
        # coordinates outside of mesh
        result[simplex == -1, :] = -1
        # convert back to 2d coords
        result = result[:, :2]
        return result

rng = np.random.default_rng(42)
for _ in range(100):
    src = rng.random((100, 2))
    dst = rng.random((100, 2))
    tform = ski.transform.PiecewiseAffineTransform()
    tform.estimate(src, dst)
    tform2 = NewPiecewiseAffineTransform()
    tform2.estimate(src, dst)
    x = rng.random((2000, 2))
    np.testing.assert_allclose(tform(x), tform2(x))
```

</p>
</details>

Additional disclaimer before approving this: I still need to wrap my head around the `einsum` magic happening here. :sweat_smile:

svepe commented:
The code before used a python loop iterating over simplices and at every iteration would i) get the simplex affine matrix, ii) get the points on the simplex and iii) apply the affine transform. The slowest bit is the loop, so the key idea is to match every point with its corresponding affine matrix and let numpy do the multiplication for all points. 

So the proposed code does 2 things:

1. Construct a stack of affine matrices (`affines` in the code) such that `affines[i]` contains the transformation matrix for the i-th point (`points[i]`). This allows us to apply the entire PiecewiseTransform with something like
```
for i in range(len(points)):
    result[i] = affines[i] * points[i]
```
but that still has the python loop.

2. So to do the multiplication in numpy we can use einsum. Multiplying a single matrix `T` with a single vector `p` with einsum can be done with `q = np.einsum("kj, j -> k". T, p)` which in math notation is equivalent to 

![image](https://github.com/scikit-image/scikit-image/assets/3459505/4d4962f6-d2a7-43b9-9a89-ea711f3dc48e)
But we want to do multiple matrix-point multiplications (`affines[i] * points[i]`) hence we do `q = np.einsum("ikj, ij -> ik", T, p)` which is equivalent to

![image](https://github.com/scikit-image/scikit-image/assets/3459505/e1bc107c-a696-4b2d-b39f-565ad30ca004)
where now `p` is a 2d matrix (`points`) and T is a 3d matrix (`affines`). Hopefully, this demystifies it a bit. Probably can be done with `numpy.dot` and a couple of extra dimensions and transpositions.

stefanv commented:
This optimization seems fine. I would recommend that we keep a version of the (original) code in a comment that explains what's happening under the hood. A text comment may also help with that. The slight change in residual is not a concern, being so small.

23pointsNorth commented:
@stefanv  Do you mean appending 

```
        # The below is equivalent to the following code, speed up by replacing manual iteration with einsum
        # for index in range(len(self._tesselation.simplices)):
            # Affine transform for triangle
            # affine = self.affines[index]
            # All coordinates within triangle
            # index_mask = simplex == index
            # out[index_mask, :] = affine(coords[index_mask, :])
```

Before the main/einsum call?

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
